Add support for enumeration tagging in the schema file.

peterwald opened the pull request:
Generate a map[EnumType]string that maps the enumeration codes to a friendly string. The string is taken from a $tag annotation added to each declaration in the schema file.

for example the following schema definition

```
using Go = import "../../jmckaskill/go-capnproto/go.capnp";

$Go.package("sample");

@0xa930f62d3f8257e3;

enum Fruits {
    apples @0 $Go.tag("Apples");
    bananas @1 $Go.tag("Bananas");
    oranges @2 $Go.tag("Oranges");
}
```

now generates 

```
package sample

// AUTO GENERATED - DO NOT EDIT

import (
    C "github.com/jmckaskill/go-capnproto"
    "unsafe"
)

type Fruits uint16

const (
    FRUITS_APPLES  Fruits = 0 // Apples
    FRUITS_BANANAS        = 1 // Bananas
    FRUITS_ORANGES        = 2 // Oranges
)

var Fruits_Tags map[Fruits]string = map[Fruits]string{
    FRUITS_APPLES:  "Apples",
    FRUITS_BANANAS: "Bananas",
    FRUITS_ORANGES: "Oranges",
}

type Fruits_List C.PointerList

func NewFruitsList(s *C.Segment, sz int) Fruits_List { return Fruits_List(s.NewUInt16List(sz)) }
func (s Fruits_List) Len() int                       { return C.UInt16List(s).Len() }
func (s Fruits_List) At(i int) Fruits                { return Fruits(C.UInt16List(s).At(i)) }
func (s Fruits_List) ToArray() []Fruits {
    return *(*[]Fruits)(unsafe.Pointer(C.UInt16List(s).ToEnumArray()))
}
```

peterwald committed:
Add support for enumeration tagging in the schema file.

Generate a map[EnumType]string that maps the enumeration codes to a friendly string. The string is taken from a $tag annotation added to each declaration in the schema file.

jmckaskill commented:
Cool. I quite like the idea, but I think it can be taken a step or two further. Is your main use case for debugging, or do you need to iterate over the set of enum options? If it's only the former I think a String() method with a switch statement would be a better approach. I'm not sure if the latter is all that useful as other parties can extend to have more values, but there are probably use cases.

Also if tag isn't specified then you get enum  { .... // <nothing here> \n which seems a bit odd. I'm actually of the opinion that we should probably always generate a tag, use $Go.tag("...") to change the name and $Go.tag("") or $Go.notag() to disable the auto tag.

peterwald commented:
Yeah. You're right. That's a better idea. I'll tweak it later this evening.
On Nov 22, 2013 6:35 PM, "James McKaskill" notifications@github.com wrote:

> Cool. I quite like the idea, but I think it can be taken a step or two
> further. Is your main use case for debugging, or do you need to iterate
> over the set of enum options? If it's only the former I think a String()
> method with a switch statement would be a better approach. I'm not sure if
> the latter is all that useful as other parties can extend to have more
> values, but there are probably use cases.
> 
> Also if tag isn't specified then you get enum { .... // \n which seems a
> bit odd. I'm actually of the opinion that we should probably always
> generate a tag, use $Go.tag("...") to change the name and $Go.tag("") or
> $Go.notag() to disable the auto tag.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/jmckaskill/go-capnproto/pull/4#issuecomment-29118108
> .

peterwald committed:
Convert enum tag lookup from a map to a generated switch statement.

Also, generate tags for each enumerant by default, and add a notag annotation for disabling the default behavior.

peterwald commented:
My use case right now is for debugging and logging and there's no need to enumerate the list. I've updated the lookup to use a case statement rather than a map so it's immutable and potentially slightly more efficient. It now automatically creates a tag for each enumerant by default, unless $Go.notag is included as an annotation (or $Go.tag("") works as well coincidentally).

**Sample File**

```
using Go = import "../../jmckaskill/go-capnproto/go.capnp";

$Go.package("sample");

@0xa930f62d3f8257e3;

enum Fruits {
    apples @0;
    bananas @1 $Go.tag("Go Bananas");
    oranges @2 $Go.notag;
    mangoes @3 $Go.tag("Mango") $Go.notag;
}

enum Vegetables {
    carrots @0 $Go.notag;
}
```

**Sample Compiled**

```
package sample

// AUTO GENERATED - DO NOT EDIT

import (
    C "github.com/jmckaskill/go-capnproto"
    "unsafe"
)

type Fruits uint16

const (
    FRUITS_APPLES  Fruits = 0
    FRUITS_BANANAS        = 1
    FRUITS_ORANGES        = 2
    FRUITS_MANGOES        = 3
)

func Fruits_Tag(c Fruits) string {
    switch c {
    case FRUITS_APPLES:
        return "apples"
    case FRUITS_BANANAS:
        return "Go Bananas"
    default:
        return ""
    }
}

type Fruits_List C.PointerList

func NewFruitsList(s *C.Segment, sz int) Fruits_List { return Fruits_List(s.NewUInt16List(sz)) }
func (s Fruits_List) Len() int                       { return C.UInt16List(s).Len() }
func (s Fruits_List) At(i int) Fruits                { return Fruits(C.UInt16List(s).At(i)) }
func (s Fruits_List) ToArray() []Fruits {
    return *(*[]Fruits)(unsafe.Pointer(C.UInt16List(s).ToEnumArray()))
}

type Vegetables uint16

const (
    VEGETABLES_CARROTS Vegetables = 0
)

func Vegetables_Tag(c Vegetables) string {
    switch c {
    default:
        return ""
    }
}

type Vegetables_List C.PointerList

func NewVegetablesList(s *C.Segment, sz int) Vegetables_List {
    return Vegetables_List(s.NewUInt16List(sz))
}
func (s Vegetables_List) Len() int            { return C.UInt16List(s).Len() }
func (s Vegetables_List) At(i int) Vegetables { return Vegetables(C.UInt16List(s).At(i)) }
func (s Vegetables_List) ToArray() []Vegetables {
    return *(*[]Vegetables)(unsafe.Pointer(C.UInt16List(s).ToEnumArray()))
}
```

jmckaskill commented:
> func Fruits_Tag(c Fruits) string
> This should be func (s Fruits_Tag) String() string instead
> That way it works with printf %v

peterwald committed:
Use idiomatic String() method to return the tag name.

Works with %v string formatting.

peterwald committed:
Added docs for enum tags.

peterwald commented:
That wraps it up for me. Let me know what you think.

jmckaskill commented:
This has been merged, but I redid the documentation as all the hard tabs had been expanded.

peterwald commented:
Thanks, and sorry about the tabs.
